server/src/main/java/com/thoughtworks/go/config/update/PartialConfigUpdateCommand.java

```diff
@@ -58,4 +75,44 @@ public CruiseConfig update(CruiseConfig cruiseConfig) {
         }
         return cruiseConfig;
     }
+
+    private boolean validateEntityForRules(PartialConfig partialConfig) {
+        //preflight check
+        if (configRepoConfig == null) {
+            return true;
+        }
+
+        if (configRepoConfig.getRules().isEmpty()) {
+            throw new InvalidPartialConfigException(partialConfig, "Configurations can not be merged as no rules are defined.");
+        }
+
+        partialConfig.getEnvironments().stream()
+                .filter(env -> !configRepoConfig.canRefer(ENVIRONMENT.getEntityType(), env.name().toString()))
+                .forEach(envThatCanNotBeReferred -> {
+                    envThatCanNotBeReferred.addError(ENVIRONMENT.getType(), format("Cannot refer environment: '%s' from the config repository.", envThatCanNotBeReferred.name()));
+                });
+
+        partialConfig.getGroups().stream()
+                .filter(pipelineGrp -> !configRepoConfig.canRefer(PIPELINE_GROUP.getEntityType(), pipelineGrp.getGroup()))
+                .forEach(pipelineGrpThatCannotBeReferred -> {
+                    pipelineGrpThatCannotBeReferred.addError(PIPELINE_GROUP.getType(), format("Cannot refer pipeline group: '%s' from the config repository.", pipelineGrpThatCannotBeReferred.getGroup()));
+                });
+
+        List<DependencyMaterialConfig> dependencyMaterialConfigs = new ArrayList<>();
+        partialConfig.getGroups().forEach((pipelineGrp) -> {
+            pipelineGrp.forEach((pipelineConfig) -> dependencyMaterialConfigs.addAll(pipelineConfig.dependencyMaterialConfigs()));
+        });
+        dependencyMaterialConfigs.stream()
+                .filter(dependencyMaterialConfig -> !doesPipelineExistInPartialConfig(partialConfig, dependencyMaterialConfig.getPipelineName()))
```

Comment:
@kritika-singh3 @arvindsv @adityasood 

**Pipeline reference rule defined on a config repository is not applicable for the pipelines coming from the same repository.**

Description:

For the following rule:
`<allow action="refer" type="pipeline">up42</pipeline>`

Following partials are coming from config repository:
* `pipeline1` (which has a pipeline-dependency `up42`)
* `pipeline2` (which has a pipeline-dependency `pipeline1`)

Even though, there is no explicit allow permission mentioned for `pipeline1` pipeline dependency, it is allowed because both the pipelines are defined in the same repository.

**Note:** We need to mention this behavior on our docs explicitly.